### PR TITLE
Feat: Add LocalVmCache for testing purposes

### PR DESCRIPTION
**Pull request by MHHukiewitz**
Problem: Programs that work on Aleph VMs and use VmCache cannot work locally

Solution: Introduce abstract base class for VmCache and implement LocalVmCache to mock VmCache functionality

### CODE REVIEW

**Comment by MHHukiewitz** [comment]
@hoh I did not include a test for VmCache, as mocking it is more complex than the code that is actually being tested. All the functionality of VmCache invokes well-known modules and functions.

File: src/aleph_client/vm/cache.py [hunk 1]
```diff
@@ -1,56 +1,121 @@
...
+
+    async def keys(self, pattern: str = "*") -> List[str]:
+        if not re.match(r"^[\w?*^\-]+$", pattern):
+            raise ValueError(
+                "Pattern may only contain letters, numbers, underscore, ?, *, ^, -"
+            )
+        all_keys = list(self._cache.keys())
+        if pattern == "*":
+            return all_keys
+        return [key for key in all_keys if re.match(pattern, key)]
```
Comment (odesenfans) [inline src/aleph_client/vm/cache.py:121]: Seems like we're mixing both regular expressions and glob expressions here. What's the actual spec?

File: src/aleph_client/vm/cache.py [hunk 2]
```diff
@@ -1,56 +1,121 @@
...
+class VmCache(BaseVmCache):
+    """Virtual Machines can use this cache to store temporary data in memory on the host."""
+
+    cache: Dict[str, bytes]
+    api_host: str
+
+    def __init__(self, session: Optional[ClientSession] = None, api_host: Optional[str] = None):
+        super().__init__(session)
+        self.cache = {}
+        self.api_host = api_host if api_host else settings.API_HOST
```
[the remark this answers is not in the document]
Comment (MHHukiewitz) [reply src/aleph_client/vm/cache.py:49]: Yup, I didn't look into how to retrieve the local API. Would that indeed be `localhost` then?

File: src/aleph_client/vm/cache.py (hunk 1 above)
Comment (hoh) [reply src/aleph_client/vm/cache.py:121]: Keys may only be composed of letters and digits that `re.match(r"^\w+$", key)`.

Comment (odesenfans) [reply src/aleph_client/vm/cache.py:121]: Yes, what bothers me here is the pattern. I see `if pattern == "*":` and then `re.match(pattern, key)` so there's something to clarify.

Comment (MHHukiewitz) [reply src/aleph_client/vm/cache.py:121]: Whoops, you're right! These are wildcard and not regular expressions: https://redis.io/docs/stack/search/reference/query_syntax/

**Comment by MHHukiewitz** [comment]
Can we still merge it though? I understand that you want to increase the coverage percentage, but in this case, I really do not see the advantage in testing the handling of HTTP requests to the VM cache.

File: src/aleph_client/vm/cache.py [hunk 3]
```diff
@@ -1,56 +1,121 @@
...
             if resp.status == 404:
                 return None
 
             resp.raise_for_status()
             return await resp.read()
 
-    async def set(self, key: str, value: Union[str, bytes]):
+    async def set(self, key: str, value: Union[str, bytes]) -> Any:
         if not re.match(r"^\w+$", key):
             raise ValueError("Key may only contain letters, numbers and underscore")
```
[the remark this answers is not in the document]
Comment (odesenfans) [reply src/aleph_client/vm/cache.py:64]: I don't get your point. This is mostly about factorizing code that we use today and reducing duplication, it wouldn't change anything to the behavior of the code.

[the remark this answers is not in the document]
Comment (MHHukiewitz) [reply src/aleph_client/vm/cache.py:64]: Ahhh, damn, focused more on the return type. I'm sorry, of course you are right :bow:

**Review by odesenfans — APPROVED** [review]

**Pull request merged**